### ISSUE TITLE
qubesdb: implement 'everything is a file' approach

### ISSUE THREAD

**Issue by adrelanos** [QubesOS/qubes-issues#1447]
https://github.com/marmarek/qubes-core-agent-linux/pull/53 implements status files for `qubes-vm-type` and made me wonder... Specifically because I was contemplating to implement something similar for `qubes-vm-persistence`.

Could qubesdb use the [everything is a file](https://en.wikipedia.org/wiki/Everything_is_a_file) approach? I.e. instead of using `qubesdb-read`, one could just use `cat /proc/qubes/vm-type`?

Perhaps this can be done for cheap? If not, then it's good to have this in mind for future development, if you did not have already and feel free to reject this one.

**Comment by ben-grande** [comment]
@adrelanos Still intend on using this method? qubesdb-read "follows" xenstore-read style, not everything is a file, but things can be read from a tool.

@marmarek Is this planned? Else I'd close as not planned.

**Comment by adrelanos** [comment]
> Still intend on using this method?

Yes. If it was available, 'everything is a file' would be the preferred method.

`qubesdb-read` isn't 100% reliable. During upgrades, when `qubesdb` is updated, `qubesdb-read` is unavailable. I don't have any examples handy, but in the past that had resulted in issues.

When I wrote this ticket, I had VMs in mind.

Nowadays, there is a similar issue in dom0. Here is a recent occurance:
https://forums.whonix.org/t/qubes-4-2-4-3-whonix-17-18-denied-sdwdate-gui-connectcheck-from-anon-whonix-to-sys-whonix/22656

The solution for that issue is `sudo systemctl restart qubes.service` or reboot. If dom0 also had an 'everything is a file' approach, the system would not be reliant on daemon status for this.

**Comment by ben-grande** [comment]
> https://forums.whonix.org/t/qubes-4-2-4-3-whonix-17-18-denied-sdwdate-gui-connectcheck-from-anon-whonix-to-sys-whonix/22656

This seems like an issue on core-admin-addon-whonix. It is missing handling some API event.

Also, `qubes(d)?.service` is different than `qubes-db(-dom0)?.service`. You can stop `qubesd` and `qubesdb` still works. If `qubesdb` is outdated because `qubesd` cannot communicate with it, both file and qubesdb-read will be outdated, so they fail together.

**Comment by ArrayBolt3** [comment]
> This seems like an issue on core-admin-addon-whonix. It is missing handling some API event.

Which event should we consider using? Right now the [missing tag fixup routine](https://github.com/QubesOS/qubes-core-admin-addon-whonix/blob/1ad1586585ac7b6a5b760f4203b10cdc995f8134/qubeswhonix/__init__.py#L253-L272) runs on `domain-load` events, which happen only during qubesd startup. Would `domain-pre-start` be a good event to also listen for?

**Comment by ben-grande** [comment]
> > This seems like an issue on core-admin-addon-whonix. It is missing handling some API event.
> 
> Which event should we consider using? Right now the [missing tag fixup routine](https://github.com/QubesOS/qubes-core-admin-addon-whonix/blob/1ad1586585ac7b6a5b760f4203b10cdc995f8134/qubeswhonix/__init__.py#L253-L272) runs on `domain-load` events, which happen only during qubesd startup. Would `domain-pre-start` be a good event to also listen for?

Looking at it again, I am not sure why the tag was not applied. I thought that `qubesd.service` restarted on upgrades, but maybe it doesn't. I don't think it should be on `domain-pre-start` but if it is safer to put there than to restart qubesd when upgrading dom0, then maybe.

**Comment by ArrayBolt3** [comment]
> Looking at it again, I am not sure why the tag was not applied.

The bug is occurring (as I understand it) when the user does something that results in a VM being assigned a qvm-feature such as `kicksecure=1` (don't immediately remember what the Whonix equivalent is called), but not yet tagged as such. This can happen when distribution-morphing Trixie to Kicksecure for instance; qubesd doesn't automatically restart when the user does that. It can also happen if an app qube's template is changed from a non-Whonix-Workstation template to a Whonix-Workstation template; qubesd also doesn't restart when that is done. Now the AppVM is a Whonix-Workstation VM, but it's not tagged as such, and it won't be tagged until the next qubesd restart.

I was suggesting domain-pre-start because that would make sure the VM always had the proper tag set before it powers on. It might be a bit extreme, but it would cover all of those scenarios I believe.

**Comment by ben-grande** [comment]
> > Looking at it again, I am not sure why the tag was not applied.
> 
> The bug is occurring (as I understand it) when the user does something that results in a VM being assigned a qvm-feature such as `kicksecure=1` (don't immediately remember what the Whonix equivalent is called), but not yet tagged as such. This can happen when distribution-morphing Trixie to Kicksecure for instance; qubesd doesn't automatically restart when the user does that. It can also happen if an app qube's template is changed from a non-Whonix-Workstation template to a Whonix-Workstation template; qubesd also doesn't restart when that is done. Now the AppVM is a Whonix-Workstation VM, but it's not tagged as such, and it won't be tagged until the next qubesd restart.

For features, `domain-feature-(set|del):kicksecure`, for template change, `property-set:template`.

**Comment by ben-grande** [comment]
Anyway, this is getting off-topic. Closing per my above comment https://github.com/QubesOS/qubes-issues/issues/1447#issuecomment-3758939809.

**Comment by andrewdavidwong** [comment]
@ben-grande: What's the [resolution](https://doc.qubes-os.org/en/latest/introduction/issue-tracking.html#resolution) (e.g., declined, not applicable)? https://github.com/QubesOS/qubes-issues/issues/1447#issuecomment-3758939809 says:

> This seems like an issue on core-admin-addon-whonix. It is missing handling some API event.

Which makes it sound like some action should be taken.

**Comment by ArrayBolt3** [comment]
@andrewdavidwong The discussion about a bug was off-topic to this feature request. The actual bug I'll file a new report for.

**Comment by ArrayBolt3** [comment]
Continuation of the discussion above: https://github.com/QubesOS/qubes-issues/issues/10645
